src/lib/uri/uri.c

```diff
@@ -7,11 +7,172 @@
 #include "parser.h"
 #include "trivia/util.h"
 
-#define XSTRNDUP(s, n) (s != NULL ? xstrndup(s, n) : NULL)
+#define XSTRNDUP(s, n)  (s != NULL ? xstrndup(s, n) : NULL)
+#define CONTAINER_INCREASE_SIZE(container, name) \
+({ \
+        __typeof__(container->name) *array = &container->name; \
+        int *count = &container->name##_##count; \
+        unsigned long elem_size = sizeof(__typeof__((*array)[0])); \
+        *array = xrealloc(*array, (*count + 1) * elem_size); \
+        __typeof__ ((*array)[0]) *elem = &((*array)[(*count)++]); \
+        memset(elem, 0, elem_size); \
+        elem; \
+})
```

Comment:
Please don't use macros like this. Better factor this code into a separate helper function and/or duplicate it.